src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs

```diff
@@ -370,126 +418,410 @@ private sealed class SendFileOperation : WriteOperation
                 SocketPal.TryCompleteSendFile(context._socket, FileHandle, ref Offset, ref Count, ref BytesTransferred, out ErrorCode);
         }
 
-        private enum QueueState
+        // In debug builds, this struct guards against:
+        // (1) Unexpected lock reentrancy, which should never happen
+        // (2) Deadlock, by setting a reasonably large timeout
+        private struct LockToken : IDisposable
         {
-            Clear = 0,
-            Set = 1,
-            Stopped = 2,
+            private object _lockObject;
```

Comment:
Nit: readonly (I particularly like readonly for lock objects, as it helps ensure we don't accidentally switch the identity of the lock)